cinema/views.py

```diff
@@ -48,11 +79,67 @@ class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
     serializer_class = MovieSessionSerializer
 
-    def get_serializer_class(self):
+    def get_queryset(self) -> QuerySet:
+        queryset = self.queryset
+
+        if self.action == "list":
+            queryset = (
+                queryset
+                .select_related("movie", "cinema_hall")
+                .annotate(
+                    tickets_available=(
+                        F("cinema_hall__rows") * F("cinema_hall__seats_in_row")
+                        - Count("tickets")
+                    )
+                )
+            )
+
+            movie = self.request.query_params.get("movie")
+            date = self.request.query_params.get("date")
+            if movie:
+                queryset = queryset.filter(movie_id=int(movie))
+            if date:
+                queryset = queryset.filter(show_time__date=date)
+
+        if self.action == "retrieve":
+            queryset = queryset.select_related("movie", "cinema_hall")
+
+        return queryset
+
+    def get_serializer_class(self) -> type[MovieSessionSerializer]:
         if self.action == "list":
             return MovieSessionListSerializer
 
         if self.action == "retrieve":
             return MovieSessionDetailSerializer
 
         return MovieSessionSerializer
+
+
+class OrderPagination(PageNumberPagination):
```

Comment:
would be better structure to put pagination in separate module and import from there